tensorboard/components/tensor_widget/slicing-control.ts

```diff
@@ -0,0 +1,335 @@
+            this.clearAllDropdowns();
+            dimControl.style.display = 'none';
+            dimInput.style.display = 'inline-block';
+          });
+
+          // Set change callback for the dim input.
+          dimInput.addEventListener('change', () => {
+            const newIndex = parseInt(dimInput.value, 10);
+            if (
+              !isFinite(newIndex) ||
+              newIndex < 0 ||
+              newIndex >= dimSize ||
+              Math.floor(dimSize) != dimSize
+            ) {
+              // Reject invalid value.
+              dimInput.value = `${this.slicingSpec.slicingDimsAndIndices[slicingDims.indexOf(i)].index}`;
+              return;
+            }
+            this.slicingSpec.slicingDimsAndIndices[
+              slicingDims.indexOf(i)
+            ].index = newIndex;
+            dimControl.textContent = `${newIndex}/${dimSize}`;
+            this.onSlicngSpecChange(this.slicingSpec);
+          });
+
+          // When defocusing (blurring) from the dim input, it changes back into
+          // a static div.
+          dimInput.addEventListener('blur', () => {
+            dimInput.style.display = 'none';
+            dimControl.style.display = 'inline-block';
+          });
+
+          this.dimControlsListenerAttached[i] = true;
+        }
+      } else {
+        if (this.slicingSpec.viewingDims[0] === i) {
+          // This is a dimension being viewed as the vertical (rows) dimension.
+          dimControl.textContent =
+            `Rows: ${this.slicingSpec.verticalRange[0]}:` +
+            `${this.slicingSpec.verticalRange[1]}/${dimSize}`;
+        } else {
+          // This is a dimension being viewed as the horizontal (columns)
+          // dimension.
+          dimControl.textContent =
+            `Cols: ${this.slicingSpec.horizontalRange[0]}:` +
+            `${this.slicingSpec.horizontalRange[1]}/${dimSize}`;
+        }
+        dimControl.classList.add('tensor-widget-dim');
+        if (!this.dimControlsListenerAttached[i]) {
+          dimControl.addEventListener('click', () => {
+            const rect = dimControl.getBoundingClientRect();
+            const top = rect.bottom;
+            const left = rect.left;
+            this.renderDropdownMenuItems(dropdown, top, left, i);
+          });
+          this.dimControlsListenerAttached[i] = true;
+        }
+      }
+    }
+  }
+
+  /**
+   * Render items in a viewing dimension's dropdown menu.
+   *
+   * The rendering is based on what dimension swappings are possible.
+   * The dropdown menu will be shown if it is non-empty after the menu items are
+   * populated.
+   *
+   * @param dropdown
+   * @param top The top coordinate of the menu.
+   * @param left The left coordinate of the menu.
+   * @param dim The dimension that the viewing dimension belongs to. E.g., `0`
+   *   for the first dimension.
+   */
+  private renderDropdownMenuItems(
+    dropdown: HTMLDivElement,
+    top: number,
```

Comment:
consider using a coord.XY pattern here, to collect the two values with similar meaning & use into a common object. 

/**
 * Represents a position in a planar coordinate frame.
 */
export interface XYCoordinates {
  x: number;
  y: number;
}

Reply:
The `top` and `left` arg names correspond nicely to the return type of [getBoundingClientRect()](https://developer.mozilla.org/en-US/docs/Web/API/Element/getBoundingClientRect). I'd argue that creating a Coordinate-like interface or struct only adds confusion, because coordinates are usually x and y, but in the HTML coordinate system, things are called left, right, top and bottom. It would be confusing if force associate left with x and top with y; it'll also be confusing if you let the members of the interface be called `left` and `top` and so forth.